tools/castle-editor/code/framedesign.pas

```diff
@@ -2026,6 +2028,24 @@ procedure TDesignFrame.PropertyGridModified(Sender: TObject);
   MarkModified;
 end;
 
+procedure TDesignFrame.PropertyEditorModified(Sender: TObject);
+var
+  Sel: TComponent;
+begin
+  Sel := SelectedComponent;
+  if Sender is TPropertyEditor then
+  begin
+    if Sel <> nil then
+      RecordUndo(Sel.Name + ': change ' + TPropertyEditor(Sender).GetName + ' to ' + TPropertyEditor(Sender).GetValue, ucHigh)
+    else
+      RecordUndo('Change ' + TPropertyEditor(Sender).GetName + ' to ' + TPropertyEditor(Sender).GetValue, ucHigh)
```

Comment:
The code that constructs this undo description is a bit different than the code doing analogous job in `PropertyGridModified`. It also doesn't contain the safeguard related to `UnreadableChars` I added there.

Can you maybe add a routine like this, and use it both inside `PropertyGridModified` and in `PropertyEditorModified`?

```pascal
function GetUndoDescription(const ComponentName, PropertyName, PropertyValue: String): String;
const
  { Unreadable chars are defined like in SReplaceChars.
    Note they include newlines, we don't want to include newlines in undo description,
    as it would make menu item look weird (actually multiline on GTK2). }
  UnreadableChars = [Low(AnsiChar) .. Pred(' '), #128 .. High(AnsiChar)];
begin
  UndoDescription := 'Change ';
  { Note that ComponentName may be '' sometimes (e.g. when Sel = nil in PropertyGridModified
    because we had multiple components selected).
    No need to check for it explicitly now, trivial addition below will do nothing in this case. }
  UndoDescription := UndoDescription + ComponentName; 
  Assert(PropertyName <> '');
  UndoDescription := UndoDescription + '.' + PropertyName;
  if CharsPos(UnreadableChars, PropertyValue) = 0 then
    UndoDescription := UndoDescription + ' to ' + SenderRowValue;
end;
```

Reply:
Another comment / question: what is the relation between `PropertyGridModified` (set as `xxx.OnModified` for all TOIPropertyGrid instances) and `PropertyEditorModified` you add?

Your code makes perfect sense if we always have one, or the other, fired. Although I see more things inside `TDesignFrame.PropertyGridModified` that should possibly go also inside `PropertyEditorModified`... 

1. Are you sure that implementing `PropertyEditorModified` is actually necessary at all? Just using `PropertyGridModified` is not enough?

2. What happens now in case of various modifications, which ones fire both `PropertyEditorModified` and `PropertyGridModified`, which ones only one of them?

    - when changing a `TCastleColorPersistent` property
    - when changing a `Name` property (note that it requires some UI update, that `PropertyGridModified` does, but `PropertyEditorModified` doesn't)
    - when changing other simple properties, like some float or string properties.

Reply:
Good point. I'll investigate it more.

Reply:
> Are you sure that implementing PropertyEditorModified is actually necessary at all? Just using PropertyGridModified is not enough?

Yes, as at least we treat `Sender` in a different way.

I did some "universalization" and improvements for Undo naming, though still we have to handle them differently. Hopefully all's fixed, but give it a close look.

Reply:
> What happens now in case of various modifications, which ones fire both PropertyEditorModified and PropertyGridModified, which ones only one of them?

It seems like often they both fire. That means that we record 2 undo events for them. However, as UndoSystem finds out that "nothing changed" and both undos have `ucHigh` - they'll follow "first come first served" principle - the first Undo is recorded, second is discarded. (Just tested and confirmed that everything works as expected)

Reply:
I would indeed expect they both fire.

When only one of them fires? I would say to document it (at `PropertyEditorModified` implementation).

Reply:
Looks like this is the pattern:
![2021-03-06-123938_1366x768_scrot](https://user-images.githubusercontent.com/10726162/110203895-1990e180-7e79-11eb-9b24-984dc4e09d84.png)

Reply:
Added this information to comment